src/Widgets/FirmwareUpdateRow.vala

```diff
@@ -53,59 +56,116 @@ public class About.Widgets.FirmwareUpdateRow : Gtk.ListBoxRow {
             column_spacing = 12,
             margin = 6
         };
-        grid.attach (icon, 0, 0, 1, 2);
+        grid.attach (image, 0, 0, 1, 2);
         grid.attach (device_name_label, 1, 0);
         grid.attach (version_label, 1, 1);
 
-        switch (device.latest_release.flag) {
-            case Firmware.ReleaseFlag.IS_UPGRADE:
-                if (device.latest_release.version == device.version) {
-                    break;
-                }
-
-                var update_button = new Gtk.Button.with_label (_("Update")) {
-                    valign = Gtk.Align.CENTER
-                };
-                update_button.clicked.connect (() => {
-                    on_update_start ();
+        var icons = device.get_icons ();
+        if (icons.data != null) {
+            image.gicon = new GLib.ThemedIcon.from_names (icons.data);
+        }
 
-                    update.begin (device, device.latest_release, (obj, res) => {
-                        update.end (res);
-                        on_update_end ();
+        FirmwareClient.get_upgrades.begin (client, device.get_id (), (obj, res) => {
+            try {
+                var upgrades = FirmwareClient.get_upgrades.end (res);
+                if (upgrades != null) {
+                    is_updatable = true;
+
+                    var release = upgrades[0];
+                    version_label.label = release.get_version ();
+
+                    var update_button = new Gtk.Button.with_label (_("Update")) {
+                        valign = Gtk.Align.CENTER
+                    };
+                    update_button.clicked.connect (() => {
+                        on_update_start ();
+
+                        update.begin (release, (obj, res) => {
+                            update.end (res);
+                            on_update_end ();
+                        });
                     });
-                });
-                grid.attach (update_button, 2, 0, 1, 2);
-                break;
-        }
+                    grid.attach (update_button, 2, 0, 1, 2);
+                }
+            } catch (Error e) {
+                debug (e.message);
+            }
+        });
+
 
         add (grid);
     }
 
-    private async void update (Firmware.Device device, Firmware.Release release) {
-        var path = yield fwupd.download_file (device, release.uri);
+    private async void update (Fwupd.Release release) {
+        var detach_caption = release.get_detach_caption ();
+        var detach_image = release.get_detach_image ();
 
-        var details = yield fwupd.get_release_details (device, path);
+        if (detach_image != null) {
+            detach_image = yield download_file (detach_image);
+        }
 
-        if (details.caption != null) {
-            if (show_details_dialog (details) == false) {
-                return;
-            }
+        if (detach_caption == null || show_details_dialog (detach_caption, detach_image) == false) {
```

Comment:
This doesn't seem right? `detach_caption` is an optional string that gives the user instructions on how to put the device into the right mode for a firmware update. If it's omitted, we don't want to cancel the update by returning here?

Reply:
Ah yup, good catch. Mixed up the logic there. Should be fixed